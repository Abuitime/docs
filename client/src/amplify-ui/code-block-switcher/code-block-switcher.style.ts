import {css} from "emotion";

export const hostStyle = css`
  display: block;
  margin-bottom: 1.5rem;

  > pre {
    display: none;
  }

  amplify-code-block {
    border-top-right-radius: 0;
    border-top-left-radius: 0;
    margin-top: 0;
  }
`;

export const activeTabStyle = css`
  font-weight: 700;
  border-bottom: 0.0625rem solid var(--code-font-color);
`;

export const tabStyle = css`
  color: var(--code-font-color);
  border-top-right-radius: 0.25rem;
  border-top-left-radius: 0.25rem;
  height: 100%;
  appearance: none;
  background-color: transparent;
  padding: 0.75rem 0.5rem;
  cursor: pointer;
  font-size: 0.8125rem;

  &:first-child {
    margin-left: 0.5rem;
  }
`;

export const tabContainerStyle = css`
  border-top-right-radius: 0.25rem;
  border-top-left-radius: 0.25rem;
  background-color: var(--code-bg-color-secondary);
`;
