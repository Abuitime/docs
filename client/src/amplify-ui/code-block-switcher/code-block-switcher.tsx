import {Component, Host, h, Prop, State} from "@stencil/core";
import {
  hostStyle,
  activeTabStyle,
  tabStyle,
  tabContainerStyle,
} from "./code-block-switcher.style";
import {css} from "emotion";

const getTabLocalStorageKey = (tabHeadingList: string): string =>
  `amplify-docs::${tabHeadingList}`;

@Component({tag: "amplify-code-block-switcher", shadow: false})
export class AmplifyCodeBlockSwitcher {
  /** the tab headings, comma-separated as a single string */
  @Prop() readonly tabHeadingList?: string;

  @State() activeChildI = 0;

  tabsHeadings: string[];

  tabLocalStorageKey: string;

  componentWillLoad() {
    if (this.tabHeadingList) {
      // get the split array of headings, to map over (for rendering buttons)
      this.tabsHeadings = this.tabHeadingList.split(",").map((e) => e.trim());
      // set the local storage key, so we can persist tab selection cross-session
      this.tabLocalStorageKey = getTabLocalStorageKey(this.tabHeadingList);
      // get from the previous session, if present
      const persistedActiveChildI =
        localStorage.getItem(this.tabLocalStorageKey) || undefined;
      // ... if persisted selection is present...
      if (typeof persistedActiveChildI === "string") {
        // set the active child in state
        this.activeChildI = parseInt(persistedActiveChildI);
      }
    }
  }

  createActiveChildISetter = (i: number) => (_event: Event): void => {
    this.activeChildI = i;
    localStorage.setItem(this.tabLocalStorageKey, String(i));
  };

  render() {
    // + 1 because CSS child indices start at 1
    const revealStyle = css`
      > pre:nth-of-type(${this.activeChildI + 1}) {
        display: initial;
      }
    `;

    return (
      <Host
        class={{
          [hostStyle]: true,
          [revealStyle]: true,
        }}
      >
        <div class={tabContainerStyle}>
          {this.tabsHeadings?.map((e, i) => {
            return (
              <button
                onClick={this.createActiveChildISetter(i)}
                class={{
                  [activeTabStyle]: this.activeChildI === i,
                  [tabStyle]: true,
                }}
                key={e}
              >
                {e}
              </button>
            );
          })}
        </div>
        <slot />
      </Host>
    );
  }
}
