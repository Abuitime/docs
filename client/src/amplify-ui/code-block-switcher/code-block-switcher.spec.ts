import {AmplifyCodeBlockSwitcher} from "./code-block-switcher";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-code-block-switcher", () => {
  describe("Component logic", () => {
    let codeBlock: AmplifyCodeBlockSwitcher;
    beforeEach(() => (codeBlock = new AmplifyCodeBlockSwitcher()));

    it("should init `tabHeadingList` as `undefined`", () =>
      expect(codeBlock.tabHeadingList).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlockSwitcher],
            html: `<amplify-code-block-switcher />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
