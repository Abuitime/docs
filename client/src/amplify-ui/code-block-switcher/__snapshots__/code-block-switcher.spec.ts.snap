// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-code-block-switcher Render logic should render 1`] = `
<amplify-code-block-switcher class="css-111qeiq css-1b97n8h">
  <!---->
  <div class="css-ywitxs"></div>
</amplify-code-block-switcher>
`;
